pkg/controller/networkpolicy/validate.go

```diff
@@ -510,16 +511,50 @@ func (a *antreaPolicyValidator) validatePeers(ingress, egress []crdv1alpha1.Rule
 		return "", true
 	}
 	for _, rule := range ingress {
+		if rule.To != nil || rule.ToService != nil {
+			return "`to` and `toService` can't be used in an ingress rule", false
+		}
+
 		msg, isValid := checkPeers(rule.From)
 		if !isValid {
 			return msg, false
 		}
 	}
 	for _, rule := range egress {
+		if rule.From != nil {
+			return "`from` can't be used in an egress rule", false
+		}
+
+		if rule.ToService != nil && (rule.To != nil || rule.Ports != nil) {
+			return fmt.Sprintf("`toService` can't be used with `to` or `ports`"), false
+		}
+
 		msg, isValid := checkPeers(rule.To)
 		if !isValid {
 			return msg, false
 		}
+
+		msg, isValid = a.validateToService(rule.ToService)
+		if !isValid {
+			return msg, false
+		}
+	}
+	return "", true
+}
+
+func (a *antreaPolicyValidator) validateToService(toService []*crdv1alpha1.ServiceReference) (string, bool) {
+	supportServiceTypeInToService := sets.NewString(
+		string(v1.ServiceTypeClusterIP),
+		string(v1.ServiceTypeNodePort),
+	)
+	for _, eachToService := range toService {
+		service, err := a.networkPolicyController.serviceLister.Services(eachToService.Namespace).Get(eachToService.Name)
+		if err != nil || service == nil {
+			return fmt.Sprintf("the Service referred in `toService` is invalid"), false
+		}
+		if !supportServiceTypeInToService.Has(string(service.Spec.Type)) {
+			return fmt.Sprintf("%s type Service is not supported in `toService`", string(service.Spec.Type)), false
+		}
```

Comment:
haven't we decided to use "weaker references" for this kind of scenario? i.e. maybe it should be possible to define the policy even if the `Service` referenced by `toService` doesn't exist yet.

cc @jianjuns

Reply:
Changed to "weaker references".